File: src/db/seed.ts
import "dotenv/config";
import { randomUUID } from "node:crypto";
import bcrypt from "bcryptjs";
import { eq } from "drizzle-orm";
import { drizzle } from "drizzle-orm/mysql2";
import mysql from "mysql2/promise";
import * as schema from "./schema";
import { SEED_PROPERTIES } from "../lib/skyra/seed";
import { propertyToInsert } from "../lib/skyra/property-mapper";

async function main() {
  const databaseUrl = process.env["DATABASE_URL"];
  if (!databaseUrl) {
    throw new Error("DATABASE_URL is required");
  }

  const adminEmail = (process.env["ADMIN_EMAIL"] ?? "admin@skyrarealty.local")
    .trim()
    .toLowerCase();
  const adminPassword = process.env["ADMIN_PASSWORD"] ?? "skyra2026";

  const connection = await mysql.createConnection(databaseUrl);
  const db = drizzle(connection, { schema, mode: "default" });

  try {
    const existingAdmins = await db
      .select()
      .from(schema.admins)
      .where(eq(schema.admins.email, adminEmail))
      .limit(1);

    if (existingAdmins.length === 0) {
      const passwordHash = await bcrypt.hash(adminPassword, 12);
      await db.insert(schema.admins).values({
        id: `admin-${randomUUID()}`,
        email: adminEmail,
        passwordHash,
      });
      console.log(`Created admin: ${adminEmail}`);
    } else {
      console.log(`Admin already exists: ${adminEmail}`);
    }

    for (const property of SEED_PROPERTIES) {
      await db
        .insert(schema.properties)
        .values(propertyToInsert(property))
        .onDuplicateKeyUpdate({
          set: {
            title: property.title,
            type: property.type,
            locality: property.locality,
            address: property.address,
            price: property.price,
            areaSqft: property.areaSqft,
            bedrooms: property.bedrooms,
            bathrooms: property.bathrooms,
            furnishing: property.furnishing,
            possessionStatus: property.possessionStatus,
            ageOfProperty: property.ageOfProperty,
            facing: property.facing,
            floorNumber: property.floorNumber,
            totalFloors: property.totalFloors,
            amenities: property.amenities,
            description: property.description,
            images: property.images,
            listedDate: property.listedDate,
            ownerContactName: property.ownerContactName,
            ownerContactPhone: property.ownerContactPhone,
          },
        });
    }

    console.log(`Upserted ${SEED_PROPERTIES.length} seed properties.`);
  } finally {
    await connection.end();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: backend/app/seed.py
"""Create tables, seed admin, and upsert demo properties."""

from __future__ import annotations

import json
import uuid
from pathlib import Path

from app.auth import hash_password
from app.config import get_settings
from app.database import Base, SessionLocal, engine
from app.models import Admin, Property

ROOT = Path(__file__).resolve().parents[1]
SEED_FILE = ROOT / "data" / "properties.json"


def seed() -> None:
    settings = get_settings()
    Base.metadata.create_all(bind=engine)
    db = SessionLocal()
    try:
        email = settings.admin_email.strip().lower()
        existing = db.query(Admin).filter(Admin.email == email).first()
        if existing is None:
            db.add(
                Admin(
                    id=f"admin-{uuid.uuid4()}",
                    email=email,
                    password_hash=hash_password(settings.admin_password),
                )
            )
            db.commit()
            print(f"Created admin: {email}")
        else:
            print(f"Admin already exists: {email}")

        if not SEED_FILE.exists():
            print(f"No seed file at {SEED_FILE}; skipping properties.")
            return

        items = json.loads(SEED_FILE.read_text(encoding="utf-8"))
        upserted = 0
        for item in items:
            prop_id = item["id"]
            row = db.get(Property, prop_id)
            if row is None:
                row = Property(id=prop_id, view_count=item.get("viewCount", 0))
                db.add(row)

            row.title = item["title"]
            row.type = item["type"]
            row.locality = item["locality"]
            row.address = item["address"]
            row.price = item["price"]
            row.area_sqft = item["areaSqft"]
            row.bedrooms = item["bedrooms"]
            row.bathrooms = item["bathrooms"]
            row.furnishing = item["furnishing"]
            row.possession_status = item["possessionStatus"]
            row.age_of_property = item["ageOfProperty"]
            row.facing = item["facing"]
            row.floor_number = item["floorNumber"]
            row.total_floors = item["totalFloors"]
            row.amenities = item["amenities"]
            row.description = item["description"]
            row.images = item["images"]
            row.listed_date = item["listedDate"]
            row.owner_contact_name = item["ownerContactName"]
            row.owner_contact_phone = item["ownerContactPhone"]
            upserted += 1

        db.commit()
        print(f"Upserted {upserted} properties.")
    finally:
        db.close()


if __name__ == "__main__":
    seed()
